docs/proposals/image-provenance.md

```diff
@@ -0,0 +1,334 @@
+    Kubernetes harder to reason about.  Also, excessive information limits cache-ability of decisions.  Caching
+    reduces latency and allows short outages of the backend to be tolerated.
+
+
+Detailed discussion in [Ensuring only images are from approved sources are run](
+https://github.com/kubernetes/kubernetes/issues/22888).
+
+# Implementation
+
+A new admission controller will be added.  That will be the only change.
+
+## Admission controller
+
+An `ImagePolicyWebhook` admission controller will be written.  The admission controller examines all pod objects which are
+created or updated.  It can either admit the pod, or reject it.  If it is rejected, the request sees a `403 FORBIDDEN`
+
+The admission controller code will go in `plugin/pkg/admission/imagepolicy`.
+
+There will be a cache of decisions in the admission controller. 
+
+If the apiserver cannot reach the webhook backend, it will log a warning and admit the pod.  The rationale here is
+that keeping the system running is most important, and people who care about image provenance deeply will
+probably separately have after-the-fact auditing of what images were run, which will catch anything that
+sneaks through this way.
+
+# Webhook Backend
+
+The admission controller code in that directory does not contain logic to make an admit/reject decision.  Instead, it extracts
+relevant fields from the Pod creation/update request and sends those fields to a Backend (which we have been loosely calling "WebHooks"
+in Kubernetes).  The request the admission controller sends to the backend is called a WebHook request to distinguish it from the
+request being admission-controlled.  The server that accepts the WebHook request from Kubernetes is called the "Backend"
+to distinguish it from the WebHook request itself, and from the API server.
+
+The whole system will work similarly to the [Authentication WebHook](
+https://github.com/kubernetes/kubernetes/pull/24902
+) or the [AuthorizationWebHook](
+https://github.com/kubernetes/kubernetes/pull/20347).
+
+The WebHook request can optionally authenticate itself to its backend using a token from a `kubeconfig` file.
+
+The WebHook request and response are JSON, and correspond to the following `go` structures:
+
+```go
+// Filename: pkg/apis/authentication.k8s.io/register.go
+package imagepolicy
+
+// ImageReview checks if the set of images in a pod are allowed.
+type ImageReview struct {
+ 	unversioned.TypeMeta
+ 
+ 	// Spec holds information about the pod being evaluated
+ 	Spec ImageReviewSpec
+ 
+ 	// Status is filled in by the backend and indicates whether the pod should be allowed.
+ 	Status ImageReviewStatus
+ }
+ 
+// ImageReviewSpec is a description of the pod creation request.
+type ImageReviewSpec struct {
+	// Containers is a list of a subset of the information in each container of the Pod being created.
+	Containers []ImageReviewContainerSpec
```

Comment:
I'm of the opinion that instead of mimicking pod layout, we should eliminate the container list altogether. The goal here is to establish a chain of trust for containers, and not to enforce pod-level policies. If you create a pod with 5 containers, then you make 5 requests to the backend.